sshd-core/src/main/java/org/apache/sshd/common/config/keys/OpenSSHKeyUtils.java

```diff
@@ -0,0 +1,150 @@
+        DEFAULT_DIGEST_HOLDER.set(ValidateUtils.checkNotNull(f, "No digest factory"));
+    }
+
+    /**
+     * @param key the public key - ignored if {@code null}
+     * @return the fingerprint or {@code null} if no key.
+     * <B>Note:</B> if exception encountered then returns the exception's simple class name
+     * @see #getFingerPrint(Factory, PublicKey)
+     */
+    public static String getFingerPrint(PublicKey key) throws Exception {
+        return getFingerPrint(getDefaultFingerPrintFactory(), key);
+    }
+
+    /**
+     * @param f   The {@link Factory} to create the {@link Digest} to use
+     * @param key the public key - ignored if {@code null}
+     * @return the fingerprint or {@code null} if no key.
+     * <B>Note:</B> if exception encountered then returns the exception's simple class name
+     * @see #getFingerPrint(Digest, PublicKey)
+     */
+    public static String getFingerPrint(Factory<? extends Digest> f, PublicKey key) throws Exception {
+        return getFingerPrint(f.create(), key);
+    }
+
+    /**
+     * @param d   The {@link Digest} to use
+     * @param key the public key - ignored if {@code null}
+     * @return the fingerprint or {@code null} if no key.
+     * @throws Exception if cannot create fingerprint.
+     */
+    public static String getFingerPrint(Digest d, PublicKey key) throws Exception {
+        if (key == null) {
+            return null;
+        }
+
+        Buffer buffer = new ByteArrayBuffer();
+        buffer.putRawPublicKey(key);
+
+        d.init();
+        d.update(buffer.array(), 0, buffer.wpos());
+
+        byte[] data = d.digest();
+
+        if ("MD5".equals(d.getAlgorithm())) {
+            return "MD5:" + BufferUtils.printHex(':', data);
+        } else {
+            return d.getAlgorithm().replace("-", "").toUpperCase() + ":" + Base64.encodeToString(data).replaceAll("=", "");
+        }
+    }
+
+    /**
+     * @param expected The expected fingerprint.
+     * @param key the public key - ignored if {@code null}
+     * @return Pair<Boolean, String> - first is success indicator, second is actual fingerprint.
+     */
+    public static Pair<Boolean, String> checkFingerPrint(String expected, PublicKey key) throws Exception {
+        if (key == null) {
+            return null;
+        }
+
+        if (expected == null) {
+            return new Pair<>(false, getFingerPrint(key));
+        }
+
+        String comps[] = expected.split(":", 2);
+        DigestFactory factory;
+        if (comps[0].length() > 2) {
+            factory = BuiltinDigests.fromString(comps[0]);
+            if (factory == null) {
+                return new Pair<>(false, getFingerPrint(key));
+            }
+        } else {
+            factory = BuiltinDigests.md5;
+            expected = "MD5:" + expected;
+        }
+
+        String fingerprint = getFingerPrint(factory, key);
+        return new Pair<>(expected.equals(fingerprint), fingerprint);
```

Comment:
For MD5 you must use _expected.equalsIgnoreCase(fingerprint)_ in case HEX is uppercase vs. lowercase.